### ISSUE TITLE
CronJobs Stall After Update - Existing Jobs Not Removed

### ISSUE THREAD

**Issue by wastrachan** [kubernetes/kubernetes#109287]
### What happened?

A `CronJob` that is scheduled to run every minute is updated. After the update, the `CronJob` stops scheduling. 

Investigation reveals the following:

1. The existing `Job` and it's associated `Pod` remain in a `completed` state, and are not removed by the scheduler 
2. The `CronJob` reports `too many missed start times: 0` 
3. The `CronJob` does not reschedule

### `kubectl get pods`
```
NAME                                                       READY   STATUS      RESTARTS   AGE
announcements-salesforce-27480650-fxf4d        0/1     Completed   0          2d18h
```

### `kubectl get jobs`
```
NAME                                                 COMPLETIONS   DURATION   AGE
announcements-salesforce-27480650        1/1           79s        2d18h
```

### `kubectl get cronjobs`
```
NAME                                        SCHEDULE                        SUSPEND   ACTIVE   LAST SCHEDULE   AGE
announcements-salesforce        * * * * *                       False     1        2d18h           160d
```

### `kubectl describe jobs announcements-salesforce-27480650`
```
Name:                     announcements-salesforce-27480650
Namespace:                default
Selector:                 controller-uid=980e1792-530c-4b58-9285-ec3845835137
Labels:                   app.kubernetes.io/instance=redacted
                          app.kubernetes.io/managed-by=Helm
                          app.kubernetes.io/name=redacted
                          app.kubernetes.io/version=v2022.03.31
                          helm.sh/chart=redacted-1.4.0
Annotations:              <none>
Controlled By:            CronJob/announcements-salesforce
Parallelism:              1
Completions:              1
Start Time:               Fri, 01 Apr 2022 14:50:00 -0400
Completed At:             Fri, 01 Apr 2022 14:51:19 -0400
Duration:                 79s
Active Deadline Seconds:  1500s
Pods Statuses:            0 Active / 1 Succeeded / 0 Failed
Pod Template:
  Labels:           controller-uid=980e1792-530c-4b58-9285-ec3845835137
                    job-name=announcements-salesforce-27480650
  Service Account:  redacted
  Containers:
   crm-announcements-salesforce:
    Image:      us.gcr.io/redacted-prod/appserver:v2022.03.31
    Port:       <none>
    Host Port:  <none>
    Args:
      python
      manage.py
      process_crm_announcements_salesforce
    Environment Variables from:
      redacted-environment  Secret  Optional: false
    Environment:           <none>
    Mounts:
      /mnt/config from config (rw)
      /mnt/pki from pki (rw)
      /mnt/prodfs from sharedfs (rw,path="sharedtmp")
      /tmp from tmp (rw)
  Volumes:
   sharedfs:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  redacted-sharedfs
    ReadOnly:   false
   pki:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  redacted-pki-files
    Optional:    false
   config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  redacted-config-files
    Optional:    false
   tmp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
Events:         <none>
```

### `kubectl describe pods announcements-salesforce-27480650-fxf4d`
```
Name:         announcements-salesforce-27480650-fxf4d
Namespace:    default
Priority:     0
Node:         gke-prod-cluster-prod-cluster-pool-354c18b4-7npk/10.48.16.51
Start Time:   Fri, 01 Apr 2022 14:50:00 -0400
Labels:       controller-uid=980e1792-530c-4b58-9285-ec3845835137
              job-name=announcements-salesforce-27480650
Annotations:  <none>
Status:       Succeeded
IP:           10.48.67.192
IPs:
  IP:           10.48.67.192
Controlled By:  Job/announcements-salesforce-27480650
Containers:
  crm-announcements-salesforce:
    Container ID:  containerd://bafb495edbde7037de4c6134f8e716b673f96a5d74e6d522d47c1e694050d978
    Image:         us.gcr.io/redacted-prod/appserver:v2022.03.31
    Image ID:      us.gcr.io/redacted-prod/appserver@sha256:37fff7b24d31742da97df6c7f7b5c732903695364183ace1651caa391394a1ed
    Port:          <none>
    Host Port:     <none>
    Args:
      python
      manage.py
      process_crm_announcements_salesforce
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Fri, 01 Apr 2022 14:50:12 -0400
      Finished:     Fri, 01 Apr 2022 14:50:58 -0400
    Ready:          False
    Restart Count:  0
    Environment Variables from:
      redacted-environment  Secret  Optional: false
    Environment:           <none>
    Mounts:
      /mnt/config from config (rw)
      /mnt/pki from pki (rw)
      /mnt/prodfs from sharedfs (rw,path="sharedtmp")
      /tmp from tmp (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-5zv2g (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             False
  ContainersReady   False
  PodScheduled      True
Volumes:
  sharedfs:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  redacted-sharedfs
    ReadOnly:   false
  pki:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  redacted-pki-files
    Optional:    false
  config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  redacted-config-files
    Optional:    false
  tmp:
    Type:       EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:  <unset>
  kube-api-access-5zv2g:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   BestEffort
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>
```

### `kubectl describe cronjobs.batch announcements-salesforce`
```
Name:                          announcements-salesforce
Namespace:                     default
Labels:                        app.kubernetes.io/instance=redacted
                               app.kubernetes.io/managed-by=Helm
                               app.kubernetes.io/name=redacted
                               app.kubernetes.io/version=v2022.04.01
                               helm.sh/chart=redacted-1.4.0
Annotations:                   meta.helm.sh/release-name: redacted
                               meta.helm.sh/release-namespace: default
Schedule:                      * * * * *
Concurrency Policy:            Forbid
Suspend:                       False
Successful Job History Limit:  1
Failed Job History Limit:      1
Starting Deadline Seconds:     <unset>
Selector:                      <unset>
Parallelism:                   <unset>
Completions:                   <unset>
Active Deadline Seconds:       1500s
Pod Template:
  Labels:           <none>
  Service Account:  redacted
  Containers:
   crm-announcements-salesforce:
    Image:      us.gcr.io/redacted-prod/appserver:v2022.04.01
    Port:       <none>
    Host Port:  <none>
    Args:
      python
      manage.py
      process_crm_announcements_salesforce
    Environment Variables from:
      redacted-environment  Secret  Optional: false
    Environment:           <none>
    Mounts:
      /mnt/config from config (rw)
      /mnt/pki from pki (rw)
      /mnt/prodfs from sharedfs (rw,path="sharedtmp")
      /tmp from tmp (rw)
  Volumes:
   sharedfs:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  redacted-sharedfs
    ReadOnly:   false
   pki:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  redacted-pki-files
    Optional:    false
   config:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  redacted-config-files
    Optional:    false
   tmp:
    Type:            EmptyDir (a temporary directory that shares a pod's lifetime)
    Medium:
    SizeLimit:       <unset>
Last Schedule Time:  Fri, 01 Apr 2022 14:50:00 -0400
Active Jobs:         announcements-salesforce-27480650
Events:
  Type     Reason              Age                     From                Message
  ----     ------              ----                    ----                -------
  Warning  TooManyMissedTimes  96s (x3913 over 2d17h)  cronjob-controller  too many missed start times: 0. Set or decrease .spec.startingDeadlineSeconds or check clock skew
```

### What did you expect to happen?

When The `CronJob` is updated: 

1. I expect the scheduler to remove the existing `Job` as it completes, as it would on a normal run. 
2. I expect that the `CronJob` will schedule a new `Job` roughly on time. 

### How can we reproduce it (as minimally and precisely as possible)?

1. Create a `CronJob` with a frequency of `* * * * *`, and a concurrency policy of `Forbid`, with a task that lasts for more than one minute. 
2. Update the `CronJob` spec (add an annotation or label)
3. Deploy the updated `CronJob` spec

### Anything else we need to know?

I suspect that the existing `Job` is still running when the `CronJob` is updated. The `CronJob` resource is replaced, and then when the `Job` completes there is some disconnect in the scheduler, as the` CronJob` that manages the `Job` has changed. 

This disconnect prevents the `Job` from being removed, and the presence of the `Job` prevents a new `Job` from being schedule (due to the concurrency policy).

In an ideal state, the scheduler would remove the completed job as normal when it completes, allowing the `CronJob` to create a new `Job` when it's scheduled to run.

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.5", GitCommit:"c285e781331a3785a7f436042c65c5641ce8a9e9", GitTreeState:"archive", BuildDate:"2022-03-17T21:14:47Z", GoVersion:"go1.18", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.9-gke.1002", GitCommit:"f87f9d952767b966e72a4bd75afea25dea187bbf", GitTreeState:"clean", BuildDate:"2022-02-25T18:12:32Z", GoVersion:"go1.16.12b7", Compiler:"gc", Platform:"linux/amd64"}

```

</details>

### Cloud provider

<details>
Google - Google Kubernetes Engine v1.21.9-gke.1002
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
LOGO=archlinux-logo

$ uname -a
Linux mithrax 5.17.1-arch1-1 #1 SMP PREEMPT Mon, 28 Mar 2022 20:55:33 +0000 x86_64 GNU/Linux
```
</details>

### Install tools

<details>
Helm 3.8.1
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by ide** [comment]
I am seeing similar symptoms (also with v1.21.9-gke.1002) where successfully completed Jobs sometimes are not removed even when `successfulJobsHistoryLimit` is set to a small number.

Like the original report in this issue, these Jobs are scheduled to run every minute and sometimes take more than a minute to complete. Similarly, the CronJob specs are continuously deployed, increasing the likelihood that the CronJob resource is updated with Jobs in progress.

Given the same symptoms and similar context, I suspect I am seeing the same issue. Sorry for not being able to provide more insight into the root cause.

**Comment by oradwell** [comment]
we also have the same issue. I previously thought it was caused by #103607 and fixed at #104799 but we're still experiencing this

**Comment by oradwell** [comment]
Found out that this is actually a duplicate of #106496 and it was fixed in #107997
I've removed "app.kubernetes.io/version" from the job objects as a workaround and it worked

@wastrachan as a workaround you can also remove "app.kubernetes.io/version" label from the job. since it changes on every deploy the bug causes the cronjob controller to lose track of the job status.
@ide you must have a similar issue please check that your job labels don't change between deploys. cronjob and pod labels are fine to change between deploys

please see the below:
```diff
--- simple-cronjob.yaml
+++ simple-cronjob.yaml
@@ -1,33 +1,32 @@
 apiVersion: batch/v1
 kind: CronJob
 metadata:
   name: sleeping-beauty
   labels:
     app.kubernetes.io/name: sleeping-beauty
     app.kubernetes.io/instance: sleeper
     app.kubernetes.io/version: v101
 spec:
   schedule: "13 * * * *"
   concurrencyPolicy: Forbid
   jobTemplate:
     metadata:
       labels:
         app.kubernetes.io/name: sleeping-beauty
         app.kubernetes.io/instance: sleeper
-        app.kubernetes.io/version: v101
     spec:
       template:
         metadata:
           labels:
             app.kubernetes.io/name: sleeping-beauty
             app.kubernetes.io/instance: sleeper
             app.kubernetes.io/version: v101
         spec:
           restartPolicy: Never
           containers:
             - image: busybox:1.34.1
               imagePullPolicy: IfNotPresent
               name: busybox
               command:
                 - sleep
                 - "3000"
```

**Comment by wastrachan** [comment]
@oradwell Great research!

#107997 was backported into several versions (including 1.22 - #108305). Presumably this would make it a candidate for a release soon. 

I do see that 1.24.0 contains this fix, but unfortunately this probably won't be available to me (on GKE) for a while. I'm going to try out your workaround, but leave this issue open for now, until I'm sure #107997 resolves this behavior.

Thanks again for the help!
